src/platform/packages/shared/kbn-esql-ast/src/definitions/utils/autocomplete/expressions/suggestionEngine.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import type { ISuggestionItem } from '../../../../commands_registry/types';
+import { getOverlapRange } from '../../shared';
+import { dispatchStates } from './positions/dispatcher';
+import { getPosition, type ExpressionPosition } from './position';
+import type {
+  ExpressionContext,
+  ExpressionContextOptions,
+  SuggestForExpressionParams,
+} from './types';
+import { isNullCheckOperator } from './utils';
+
+const WHITESPACE_REGEX = /\s/;
+const LAST_WORD_BOUNDARY_REGEX = /\b\w(?=\w*$)/;
+
+/** Coordinates position detection, handler selection, and range attachment */
+export async function suggestForExpression(
+  params: SuggestForExpressionParams
+): Promise<ISuggestionItem[]> {
+  const baseCtx = buildContext(params);
+  const position: ExpressionPosition = getPosition(baseCtx.innerText, baseCtx.expressionRoot);
+
+  const ctx = { ...baseCtx, position };
+  const suggestions = await dispatchStates(ctx, position);
+
+  attachRanges(ctx, suggestions);
+
+  return suggestions;
+}
+
+/** Derives innerText and option flags from the incoming params.*/
+export function buildContext(params: SuggestForExpressionParams): ExpressionContext {
```

Comment:
This function adds a parameter at the end. I might even remove it now that I think about it. 
It's still here for a historical reason: in the early versions of refctoring, I had more things to normalize.